staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -82,29 +74,92 @@ func init() {
+	active map[string]*activeETCD3Client
+}
+
+func newETCD3ClientCache() *etcd3ClientCache {
+	return &etcd3ClientCache{active: make(map[string]*activeETCD3Client)}
+}
+
+func etcd3ClientCacheKey(tc storagebackend.TransportConfig) string {
+	return fmt.Sprintf("%v", tc) // gives: {[server1 server2] keyFile certFile caFile}
+}
+
+// For returns a compacted etcd v3 client for the supplied transport config. One
+// client and compactor are started for each transport. Callers must call the
+// returned destroy function once they are done with the client.
+func (e *etcd3ClientCache) For(tc storagebackend.TransportConfig, compactionInterval, metricPollInterval time.Duration) (*clientv3.Client, error) {
+	e.mx.Lock()
+	defer e.mx.Unlock()
+
+	k := etcd3ClientCacheKey(tc)
+	if a, ok := e.active[k]; ok {
+		// We want to compact and poll at the shortest interval specified by any caller.
+		if a.compactionInterval > compactionInterval || a.metricPollInterval > metricPollInterval {
+			a.cancel()
+			ctx, cancel := context.WithCancel(context.Background())
+			etcd3.StartCompactor(ctx, a.client, compactionInterval)
+			etcd3.StartDBSizeMonitor(ctx, a.client, metricPollInterval)
+			a.compactionInterval = compactionInterval
+			a.metricPollInterval = metricPollInterval
+			a.cancel = cancel
+		}
+		a.refs++
+		return a.client, nil
+	}
 
-	l, err := logutil.CreateDefaultZapLogger(etcdClientDebugLevel())
+	c, err := newETCD3Client(tc)
 	if err != nil {
-		l = zap.NewNop()
+		return nil, fmt.Errorf("cannot create new etcd3 client: %w", err)
 	}
-	etcd3ClientLogger = l.Named("etcd-client")
+
+	// Decorate the KV instance so we can track etcd latency per request.
+	// TODO(negz): Does it hurt that this will track latency for compaction?
+	c.KV = etcd3.NewETCDLatencyTracker(c.KV)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	etcd3.StartCompactor(ctx, c, compactionInterval)
+	etcd3.StartDBSizeMonitor(ctx, c, metricPollInterval)
+
+	e.active[k] = &activeETCD3Client{
+		client:             c,
+		cancel:             cancel,
+		refs:               1,
+		compactionInterval: compactionInterval,
+		metricPollInterval: metricPollInterval,
+	}
+	return c, nil
 }
 
-// etcdClientDebugLevel translates ETCD_CLIENT_DEBUG into zap log level.
-// NOTE(negz): This is a copy of a private etcd client function:
-// https://github.com/etcd-io/etcd/blob/v3.5.4/client/v3/logger.go#L47
-func etcdClientDebugLevel() zapcore.Level {
-	envLevel := os.Getenv("ETCD_CLIENT_DEBUG")
-	if envLevel == "" || envLevel == "true" {
-		return zapcore.InfoLevel
+func (e *etcd3ClientCache) Done(tc storagebackend.TransportConfig) {
+	e.mx.Lock()
+	defer e.mx.Unlock()
+
+	k := etcd3ClientCacheKey(tc)
+	if _, ok := e.active[k]; !ok {
+		return
 	}
-	var l zapcore.Level
-	if err := l.Set(envLevel); err == nil {
-		log.Printf("Deprecated env ETCD_CLIENT_DEBUG value. Using default level: 'info'")
-		return zapcore.InfoLevel
+
+	e.active[k].refs--
+	if e.active[k].refs > 0 {
+		return
```

Comment:
ic, so the shutdown process implemented is "the last one closes everything"
I wonder if it should it be "closes everything just once", i.e., any call will cancel everything and subsequent calls will be noop

Reply:
We need to leave the client running until all callers are done with it though, right? My understanding is one storage instance calling `Done` doesn't mean there aren't others that could still be using the same client.

Reply:
yeah, I got iy wrong, sorry https://github.com/kubernetes/kubernetes/pull/111559/files/3e22ef6f98687fe24fb094b12d0cd37075dc2b08#r936319784